File: python/src/revenant/network/legacy_tls.py
# pyright: reportUnknownMemberType=false, reportUnknownVariableType=false, reportUnknownArgumentType=false
"""Legacy TLS transport using tlslite-ng.

Some CoSign appliances (notably EKENG's ca.gov.am) require TLSv1.0 with
RC4-MD5 -- a cipher suite removed from OpenSSL 3.x.  This module provides
a raw HTTP-over-TLS implementation using tlslite-ng's pure-Python TLS stack.

Used internally by ``transport.py`` when a host requires legacy TLS.
"""

from __future__ import annotations

import logging
import socket
import time
from typing import Literal
from urllib.parse import urlparse

from tlslite import HandshakeSettings, TLSConnection
from tlslite.errors import BaseTLSException

from ..constants import (
    BYTES_PER_MB,
    DEFAULT_TIMEOUT_LEGACY_TLS,
    MAX_RESPONSE_SIZE,
    RECV_BUFFER_SIZE,
)
from ..errors import RevenantError, TLSError

_logger = logging.getLogger(__name__)

# Standard HTTP(S) ports -- used to decide Host header format
_STANDARD_PORTS = (80, 443)


def make_legacy_settings() -> HandshakeSettings:
    """Build TLS 1.0 + RC4 handshake settings for legacy appliances."""
    settings = HandshakeSettings()
    settings.cipherNames = ["rc4"]
    settings.macNames = ["md5", "sha"]
    settings.minVersion = (3, 1)  # TLS 1.0
    # maxVersion left at default (3,4) -- server negotiates down to TLS 1.0
    return settings


def _parse_status_code(status_line: str) -> int:
    """Parse HTTP status code from a status line like 'HTTP/1.0 200 OK'.

    Raises:
        TLSError: If the status line cannot be parsed.
    """
    parts = status_line.split(maxsplit=2)
    if len(parts) >= 2:
        try:
            return int(parts[1])
        except ValueError:
            pass
    raise TLSError(f"Cannot parse HTTP status line: {status_line!r}")


def legacy_request(
    method: Literal["GET", "POST"],
    url: str,
    body: bytes | None = None,
    headers: dict[str, str] | None = None,
    timeout: int = DEFAULT_TIMEOUT_LEGACY_TLS,
) -> bytes:
    """
    Send an HTTP request over legacy TLS (TLS 1.0 + RC4) using tlslite-ng.

    Used for appliances whose TLS stack requires RC4-MD5 -- a cipher
    removed from OpenSSL 3.x.

    Args:
        method: HTTP method (GET, POST).
        url: Target URL.
        body: Request body (for POST).
        headers: Additional HTTP headers.
        timeout: Socket timeout in seconds.

    Returns:
        Response body as bytes.

    Raises:
        TLSError: On connection or TLS handshake failures.
    """
    parsed = urlparse(url)
    host = parsed.hostname
    port = parsed.port or 443
    path = parsed.path or "/"
    if parsed.query:
        path = f"{path}?{parsed.query}"

    if not host:
        raise TLSError(f"Invalid URL: {url}")

    _logger.debug("%s %s (host=%s, port=%d, timeout=%ds)", method, url, host, port, timeout)

    try:
        sock = socket.create_connection((host, port), timeout=timeout)
    except TimeoutError as exc:
        raise TLSError(
            f"Connection timed out after {timeout}s. Is the server reachable?",
            retryable=True,
        ) from exc
    except OSError as exc:
        raise TLSError(
            f"Cannot connect to {host}:{port}: {exc}",
            retryable=True,
        ) from exc

    try:
        tls = TLSConnection(sock)
        # IIS 5.1 closes TCP without sending TLS close_notify alert.
        # Without this flag, recv() raises TLSAbruptCloseError on last read.
        tls.ignoreAbruptClose = True
        # NOTE: tlslite-ng does NOT verify the server certificate by default.
        # This is a known limitation for legacy TLS connections.  The target
        # servers (EKENG ca.gov.am) are accessed over a government intranet
        # and require RC4-MD5 which precludes modern certificate validation.
        tls.handshakeClientCert(settings=make_legacy_settings())
        _logger.warning(
            "Using legacy TLS (TLS 1.0 + RC4) for %s:%d. "
            "This cipher suite is deprecated and only used for backward compatibility.",
            host,
            port,
        )

        response_body = _send_and_receive(tls, method, host, port, path, body, headers, timeout)
    except (TLSError, RevenantError):
        raise
    except TimeoutError as exc:
        raise TLSError(
            f"Connection timed out after {timeout}s. Is the server reachable?",
            retryable=True,
        ) from exc
    except (OSError, ConnectionError) as exc:
        raise TLSError(
            f"Connection to {host}:{port} failed: {exc}",
            retryable=True,
        ) from exc
    except BaseTLSException as exc:
        raise TLSError(f"TLS error with {host}:{port}: {exc}") from exc
    else:
        return response_body
    finally:
        sock.close()


def _validate_header_value(name: str, value: str) -> str:
    """Reject header values containing CR/LF to prevent HTTP header injection (CWE-113)."""
    if "\r" in value or "\n" in value:
        raise TLSError(f"HTTP header '{name}' contains invalid CR/LF characters")
    return value


def _send_and_receive(
    tls: TLSConnection,
    method: str,
    host: str,
    port: int,
    path: str,
    body: bytes | None,
    headers: dict[str, str] | None,
    timeout: int = DEFAULT_TIMEOUT_LEGACY_TLS,
) -> bytes:
    """Build HTTP request, send it, and parse the response.

    Args:
        tls: Established TLS connection.
        method: HTTP method.
        host: Target hostname.
        port: Target port.
        path: Request path (including query string).
        body: Request body (for POST).
        headers: Additional HTTP headers.
        timeout: Wall-clock timeout for the entire response read.

    Returns:
        Response body as bytes.

    Raises:
        TLSError: On protocol or size-limit violations.
        RevenantError: On non-2xx HTTP status.
    """
    # Build raw HTTP/1.0 request.  HTTP/1.0 avoids chunked encoding
    # and Expect:100-continue issues with IIS 5.1 on the EKENG appliance.
    host_header = f"{host}:{port}" if port not in _STANDARD_PORTS else host
    all_headers: dict[str, str] = {"Host": host_header, "Connection": "close"}
    if body is not None:
        all_headers["Content-Length"] = str(len(body))
    if headers:
        all_headers.update(headers)

    # Validate all header values to prevent CRLF injection (CWE-113).
    # Raw HTTP construction makes this critical -- unlike urllib which
    # validates internally, we build request bytes manually.
    header_lines = "".join(
        f"{k}: {_validate_header_value(k, v)}\r\n" for k, v in all_headers.items()
    )
    request_bytes = f"{method} {path} HTTP/1.0\r\n{header_lines}\r\n".encode()
    if body:
        request_bytes += body

    tls.sendall(request_bytes)

    # Read full response (server closes connection after HTTP/1.0)
    raw = _read_response(tls, host, port, timeout)

    # Split headers from body
    header_end = raw.find(b"\r\n\r\n")
    if header_end == -1:
        raise TLSError(f"Invalid HTTP response from {host}:{port}")

    response_body = raw[header_end + 4 :]
    status_line = raw[: raw.find(b"\r\n")].decode("utf-8", errors="replace")
    _logger.debug("%s -> %s (%d bytes)", method, status_line, len(response_body))

    # Validate HTTP status code
    status_code = _parse_status_code(status_line)
    if status_code < 200 or status_code >= 300:
        raise RevenantError(f"HTTP {status_code} from {host}:{port}: {status_line}")

    return response_body


def _read_response(
    tls: TLSConnection, host: str, port: int, timeout: int = DEFAULT_TIMEOUT_LEGACY_TLS
) -> bytes:
    """Read full TLS response with size limit and wall-clock timeout."""
    chunks: list[bytes] = []
    total_size = 0
    deadline = time.monotonic() + timeout
    while True:
        if time.monotonic() > deadline:
            raise TLSError(
                f"Response from {host}:{port} exceeded {timeout}s wall-clock timeout",
                retryable=True,
            )
        chunk = tls.recv(RECV_BUFFER_SIZE)
        if not chunk:
            break
        total_size += len(chunk)
        if total_size > MAX_RESPONSE_SIZE:
            raise TLSError(
                f"Response from {host}:{port} exceeds {MAX_RESPONSE_SIZE // BYTES_PER_MB} MB limit"
            )
        chunks.append(chunk)
    return b"".join(chunks)

File: typescript/src/network/legacy-tls.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Legacy TLS transport using node-forge.
 *
 * Some CoSign appliances (notably EKENG's ca.gov.am) require TLSv1.0 with
 * RC4-MD5 -- a cipher suite removed from OpenSSL 3.x. This module provides
 * a raw HTTP-over-TLS implementation using node-forge's pure-JS TLS stack.
 */

import * as net from "node:net";
import * as forgeNamespace from "node-forge";
import "./rc4-cipher-suite.js"; // Register RC4 cipher suites with node-forge TLS

// node-forge is CJS. When loaded via dynamic import() chain, named exports
// may be undefined — only the default export works. Handle both cases.
const forge =
  (forgeNamespace as Record<string, unknown>).tls !== undefined
    ? forgeNamespace
    : ((forgeNamespace as Record<string, unknown>).default as typeof forgeNamespace);

import { BYTES_PER_MB, DEFAULT_TIMEOUT_LEGACY_TLS, MAX_RESPONSE_SIZE } from "../constants.js";
import { isNodeError, RevenantError, TLSError } from "../errors.js";
import { logger } from "../logger.js";
import { checkServerPin } from "./tls-pinning.js";

const STANDARD_PORTS = new Set([80, 443]);

function parseStatusCode(statusLine: string): number {
  const parts = statusLine.split(/\s+/, 3);
  if (parts.length >= 2) {
    const statusPart = parts[1];
    if (statusPart !== undefined) {
      const code = parseInt(statusPart, 10);
      if (!Number.isNaN(code)) return code;
    }
  }
  throw new TLSError(`Cannot parse HTTP status line: ${JSON.stringify(statusLine)}`);
}

/**
 * Re-encode a certificate's public key as a DER SubjectPublicKeyInfo.
 *
 * node-forge exposes the parsed key rather than the original bytes, so this
 * writes the structure back out. SubjectPublicKeyInfo has one valid DER
 * encoding, which is why the digest still matches one taken from the wire.
 */
function subjectPublicKeyInfoDer(cert: forgeNamespace.pki.Certificate): Uint8Array {
  const spki = forge.pki.publicKeyToAsn1(cert.publicKey);
  const der = forge.asn1.toDer(spki).getBytes();
  return Uint8Array.from(der, (ch) => ch.charCodeAt(0));
}

function validateHeaderValue(name: string, value: string): string {
  if (value.includes("\r") || value.includes("\n")) {
    throw new TLSError(`HTTP header '${name}' contains invalid CR/LF characters`);
  }
  return value;
}

export async function legacyRequest(
  method: "GET" | "POST",
  url: string,
  options: {
    body?: Uint8Array;
    headers?: Record<string, string>;
    timeout?: number;
    /**
     * Accepted server keys. Checked as soon as the certificate arrives and
     * before anything is sent; an empty list is refused rather than treated
     * as "any key".
     */
    pins: readonly string[];
  },
): Promise<Uint8Array> {
  const parsed = new URL(url);
  const host = parsed.hostname;
  const port = parsed.port ? parseInt(parsed.port, 10) : 443;
  let path = parsed.pathname || "/";
  if (parsed.search) {
    path = `${path}${parsed.search}`;
  }

  if (!host) {
    throw new TLSError(`Invalid URL: ${url}`);
  }

  const timeout = options.timeout ?? DEFAULT_TIMEOUT_LEGACY_TLS;
  const body = options.body;
  const extraHeaders = options.headers;
  const pins = options.pins;

  return new Promise<Uint8Array>((resolve, reject) => {
    // The pin check runs inside node-forge's verify callback, which can only
    // answer yes or no. The reason is kept here so the rejection says why.
    let pinFailure: Error | null = null;

    const timer = setTimeout(() => {
      socket.destroy();
      reject(
        new TLSError(`Connection timed out after ${timeout}s. Is the server reachable?`, {
          retryable: true,
        }),
      );
    }, timeout * 1000);

    const socket = net.createConnection({ host, port }, () => {
      // Socket connected, start TLS handshake
      const tls = forge.tls.createConnection({
        server: false,
        verify: (_connection, _verified, depth, certs) => {
          // The chain cannot be validated -- these appliances present a
          // self-signed certificate no authority vouches for -- so the
          // pinned key stands in for the whole of PKI. certs[0] is the
          // end-entity certificate at every depth; check it once.
          if (depth !== 0) {
            return true;
          }
          const leaf = certs[0];
          if (leaf === undefined) {
            pinFailure = new TLSError(
              `${host}:${port} completed a handshake without presenting a certificate`,
            );
            return { message: pinFailure.message };
          }
          try {
            checkServerPin(subjectPublicKeyInfoDer(leaf), pins, host, port);
          } catch (err) {
            pinFailure = err instanceof Error ? err : new TLSError(String(err));
            return { message: pinFailure.message };
          }
          return true;
        },
        connected: (connection) => {
          // TLS handshake complete, send HTTP request
          logger.info(
            `Legacy TLS (TLS 1.0 + RC4) established with ${host}:${port} against a pinned key`,
          );

          const hostHeader = STANDARD_PORTS.has(port) ? host : `${host}:${port}`;
          const allHeaders: Record<string, string> = {
            Host: hostHeader,
            Connection: "close",
          };
          if (body) {
            allHeaders["Content-Length"] = String(body.length);
          }
          if (extraHeaders) {
            Object.assign(allHeaders, extraHeaders);
          }

          const headerLines = Object.entries(allHeaders)
            .map(([k, v]) => `${k}: ${validateHeaderValue(k, v)}`)
            .join("\r\n");
          const requestStr = `${method} ${path} HTTP/1.0\r\n${headerLines}\r\n\r\n`;
          connection.prepare(requestStr);
          if (body) {
            // Convert Uint8Array to binary string without spread operator
            // (spread causes RangeError for large arrays)
            let binaryStr = "";
            for (let i = 0; i < body.length; i++) {
              binaryStr += String.fromCharCode(body[i] ?? 0);
            }
            const bodyBuf = forge.util.createBuffer(binaryStr);
            connection.prepare(bodyBuf.getBytes());
          }
        },
        tlsDataReady: (connection) => {
          const data = connection.tlsData.getBytes();
          socket.write(data, "binary");
        },
        dataReady: (connection) => {
          responseBuffer += connection.data.getBytes();
          const totalSize = responseBuffer.length;
          if (totalSize > MAX_RESPONSE_SIZE) {
            clearTimeout(timer);
            socket.destroy();
            reject(
              new TLSError(
                `Response from ${host}:${port} exceeds ${MAX_RESPONSE_SIZE / BYTES_PER_MB} MB limit`,
              ),
            );
          }
        },
        closed: () => {
          clearTimeout(timer);
          try {
            const result = parseResponse(responseBuffer, host, port);
            resolve(result);
          } catch (err) {
            reject(err);
          }
        },
        error: (_connection, error) => {
          clearTimeout(timer);
          socket.destroy();
          // A rejected pin surfaces here as a generic bad-certificate alert;
          // report what actually went wrong instead.
          reject(pinFailure ?? new TLSError(`TLS error with ${host}:${port}: ${error.message}`));
        },
      });

      let responseBuffer = "";

      socket.on("data", (data) => {
        tls.process(data.toString("binary"));
      });

      socket.on("end", () => {
        tls.close();
      });

      // Initiate TLS ClientHello -- must be called after creating the
      // connection and registering the data handler.
      tls.handshake();
    });

    socket.on("error", (err) => {
      clearTimeout(timer);
      if (isNodeError(err, "ETIMEDOUT")) {
        reject(
          new TLSError(`Connection timed out after ${timeout}s. Is the server reachable?`, {
            retryable: true,
          }),
        );
      } else {
        reject(
          new TLSError(`Cannot connect to ${host}:${port}: ${err.message}`, {
            retryable: true,
          }),
        );
      }
    });

    socket.setTimeout(timeout * 1000, () => {
      socket.destroy();
      reject(
        new TLSError(`Connection timed out after ${timeout}s. Is the server reachable?`, {
          retryable: true,
        }),
      );
    });
  });
}

function parseResponse(raw: string, host: string, port: number): Uint8Array {
  const headerEnd = raw.indexOf("\r\n\r\n");
  if (headerEnd === -1) {
    throw new TLSError(`Invalid HTTP response from ${host}:${port}`);
  }

  const responseBody = raw.slice(headerEnd + 4);
  const statusLine = raw.slice(0, raw.indexOf("\r\n"));

  const statusCode = parseStatusCode(statusLine);
  if (statusCode < 200 || statusCode >= 300) {
    throw new RevenantError(`HTTP ${statusCode} from ${host}:${port}: ${statusLine}`);
  }

  // Convert binary string to Uint8Array
  const bytes = new Uint8Array(responseBody.length);
  for (let i = 0; i < responseBody.length; i++) {
    bytes[i] = responseBody.charCodeAt(i);
  }
  return bytes;
}
